src/tokenizer.rs

```diff
@@ -2227,6 +2260,43 @@ mod tests {
         compare(expected, tokens);
     }
 
+    #[test]
+    fn tokenize_numeric_literal_underscore() {
+        let dialect = GenericDialect {};
+        let sql = String::from("SELECT 10_000");
+        let mut tokenizer = Tokenizer::new(&dialect, &sql);
+        let tokens = tokenizer.tokenize().unwrap();
+        let expected = vec![
+            Token::make_keyword("SELECT"),
+            Token::Whitespace(Whitespace::Space),
+            Token::Number("10".to_string(), false),
+            Token::make_word("_000", None),
+        ];
+        compare(expected, tokens);
+
+        let dialect = ClickHouseDialect {};
+        let sql = String::from("SELECT 10_000, _10_000, 10_00_, 10___0");
+        let mut tokenizer = Tokenizer::new(&dialect, &sql);
+        let tokens = tokenizer.tokenize().unwrap();
+        let expected = vec![
+            Token::make_keyword("SELECT"),
+            Token::Whitespace(Whitespace::Space),
+            Token::Number("10_000".to_string(), false),
+            Token::Comma,
+            Token::Whitespace(Whitespace::Space),
+            Token::make_word("_10_000", None), // leading underscore tokenizes as a word (parsed as column identifier)
+            Token::Comma,
+            Token::Whitespace(Whitespace::Space),
+            Token::Number("10_00".to_string(), false),
+            Token::make_word("_", None), // trailing underscores tokenizes as a word (syntax error in some dialects)
+            Token::Comma,
+            Token::Whitespace(Whitespace::Space),
+            Token::Number("10".to_string(), false),
+            Token::make_word("___0", None), // multiple underscores tokenizes as a word (syntax error in some dialects)
+        ];
+        compare(expected, tokens);
```

Comment:
There's a utility `.tokenizes_to(...)` that might be useful for tests like this :)